src/theory/strings/arith_entail.cpp

```diff
@@ -100,6 +101,116 @@ Node ArithEntail::rewriteArith(Node a)
   return an;
 }
 
+Node ArithEntail::normalizeGeq(const Node& n) const
+{
+  NodeManager* nm = NodeManager::currentNM();
+  if (n.getNumChildren() != 2 || !n[0].getType().isInteger()
+      || !n[1].getType().isInteger())
+  {
+    return Node::null();
+  }
+  switch (n.getKind())
+  {
+    case Kind::GEQ: return n;
+    case Kind::LEQ: return nm->mkNode(Kind::GEQ, n[1], n[0]);
+    case Kind::LT:
+      return nm->mkNode(
+          Kind::GEQ,
+          n[1],
+          nm->mkNode(Kind::ADD, n[0], nm->mkConstInt(Rational(1))));
+    case Kind::GT:
+      return nm->mkNode(
+          Kind::GEQ,
+          n[0],
+          nm->mkNode(Kind::ADD, n[1], nm->mkConstInt(Rational(1))));
+    default: break;
+  }
+  return Node::null();
+}
+
+Node ArithEntail::rewriteLengthIntro(const Node& n,
+                                     TConvProofGenerator* pg) const
+{
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<TNode, Node> visited;
+  std::unordered_map<TNode, Node>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
```

Comment:
Is this loop a performance bottleneck? It's quite complicated for what it is doing... I would just split it into two loops: the first one for getting a stack of children and the second one for processing them.

Reply:
This method is potentially a bottleneck.

The problem is that "processing" them means replacing the subterm at that point. One would actually need 2 traversals and 1 loop to process (a traversal to collect, process, another traversal to construct the term). So I'd prefer keeping a single traversal for performance.